examples/server-alarms-and-conditions.py

```diff
@@ -0,0 +1,107 @@
+# coding: utf-8
+
+from aioconsole import ainput
+import logging
+
+import asyncio
+
+from asyncua import Server, ua
+logging.basicConfig(level=logging.WARNING)
+
+class OpcUaServer(object):
+
+    def __init__(self, endpoint):
+        self.server = Server()
+        self.server.set_endpoint(endpoint)
+        self.server.set_server_name('Alarms and Conditions Test Server')
+        self.server.application_type = ua.ApplicationType.Server
+        self.server.iserver.bind_condition_methods = 500
+        self.con_gen = None
+        self.alarm_gen = None
+
+    async def __aenter__(self):
+        await self.init()
+        await self.server.start()
+        return self
+
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        await self.server.stop()
+
+    async def init(self):
+        await self.server.init()
+        uri = "http://examples.freeopcua.github.io"
+        idx = await self.server.register_namespace(uri)
+        objects = self.server.get_objects_node()
+
+        noti_node = await objects.add_object(idx, 'NotifierObject')
+
+        con_obj = await noti_node.add_object(idx, "ConditionObject")
+        condition = self.server.get_node(ua.NodeId(2830))
+        self.con_gen = await self.server.get_event_generator(condition, self.server.nodes.server,
+                                                             notifier_path=[ua.ObjectIds.Server])
+        self.con_gen.event.add_property('NodeId', con_obj.nodeid, ua.VariantType.NodeId)
+
+        alarm_obj = await noti_node.add_object(idx, "AlarmObject")
+        alarm = self.server.get_node(ua.NodeId(10637))
+        self.alarm_gen = await self.server.get_event_generator(alarm, self.server.nodes.server,
+                                                               notifier_path=[ua.ObjectIds.Server])
+
+    async def generate_condition(self, retain):
+        self.con_gen.event.ConditionName = 'Example Condition'
+        self.con_gen.event.Message = ua.LocalizedText("Some Message")
+        self.con_gen.event.Severity = 500
+        self.con_gen.event.BranchId = ua.NodeId(0)
+        if retain == 1:
+            self.con_gen.event.Retain = True
+        else:
+            self.con_gen.event.Retain = False
+        await self.con_gen.trigger()
+
+    async def generate_alarm(self, active):
+        self.alarm_gen.event.ConditionName = 'Example Alarm1'
+        self.alarm_gen.event.Message = ua.LocalizedText("hello from python")
```

Comment:
Add missing ConditionId:
```suggestion
    async def generate_alarm(self, active):
        self.alarm_gen.event.NodeId = ua.NodeId("Id222_EA") # Unique ConditionId
        self.alarm_gen.event.ConditionName = 'Example Alarm1'
        self.alarm_gen.event.Message = ua.LocalizedText("hello from python")
```

Reply:
`self.alarm_gen.event.NodeId = ua.NodeId("Id222_EA") # Unique ConditionId`
this ConditionId topic confuses me if its not like:
a condition is usually a instance of a condition type in the addressspace! so why not use the instance's nodeid?

Reply:
The ConditionId is the identifier of an alarm/condition. A device, or machine can have multiple existing alarms at the same time. How to distinguish them from each other? The client gets an ConditionEvent with Retain=False. How does the client know which of the 10 existing alarms/conditions is not existing anymore? You need an Id for that. It is like the primary key in a database. For that purpose there is the ConditionId, and for for that reason it also got its name.  

It is completely up to the server after which rules he creates the ConditionId, as long as each existing condition has a unique ConditionId. You could just use GUIDs. Or for example on a machine tool you could create a combined string using the alarm number and some parameters, as long as you know, that combination can exist only once at a time.

Why not use the instance's nodeid? Because something like that does not exist. A condition/alarm is an event. An event only exists temporarily and is gone. An event does not exist in the adresspace like a node. There is only a node in the addresspace where you can register for the events. An event is not a node and therefore there is no "instance's nodeid".

Because there is no "instance's nodeid", they added something with a similar meaning and called it ConditionId (of type NodeId). How to fill it, is up to the server.

Reply:
> Why not use the instance's nodeid? Because something like that does not exist. A condition/alarm is an event. An event only exists temporarily and is gone.

this is not true, conditions are more then just events!

for example a implementation of a levelalarm which is a subtype of ConditionType ->AcknowledgeableConditionType -> AlarmConditionType -> LimitAlarmType -> ExclusiveLimitAlarmType -> "ExclusiveLevelAlarmType":
![Unbenannt](https://user-images.githubusercontent.com/56362817/102725539-b8e11680-4317-11eb-8edd-8b3081fdbe0f.PNG)
mylevelalarm has the node id "ns=5;s=MyLevel.Alarm" which is unique
![Unbenannt1](https://user-images.githubusercontent.com/56362817/102725541-bbdc0700-4317-11eb-819e-3d57a5d50810.PNG)
![Unbenannt2](https://user-images.githubusercontent.com/56362817/102725543-bd0d3400-4317-11eb-8c99-81dc03bebb26.PNG)

> It is completely up to the server after which rules he creates the ConditionId, as long as each existing condition has a unique ConditionId. You could just use GUIDs. Or for example on a machine tool you could create a combined string using the alarm number and some parameters, as long as you know, that combination can exist only once at a time.

there might be no rule but it is a good idea to look on the opc foundation defined conditions and there implementation  / usage!

you can also look into the "Unified Automation Cpp Demo Server" the same game like the prosys demo server!

Reply:
Hi,
I know real world servers where it is done as I described. And you're right there seem to be implementations as you describe as well.

It seems there are two common approaches, the servers I know use approach 2.

1. Create a node for each possible Condition/Alarm instance in the address space. Maybe that makes more sense, if you have only a couple of different Condition/Alarms or you have only a few very specialised Alarms like the ExclusiveLevelAlarm.

2. Do not create a node for each possible Condition/Alarm instance in the address space. Manage the conditions in a server internal data structure. Maybe that makes more sense, when you have thousands of different alarms in your system. Otherwise you would need to create thousands of nodes.

It is interesting, the specification also always describes two ways to do it and has examples for both approaches.
https://reference.opcfoundation.org/v104/Core/docs/Part9/6.3/
https://reference.opcfoundation.org/v104/Core/docs/Part9/B.2/

In January I can talk to a member of the opc ua technical council and get more clarification. That is the really hard part of OPC UA. They allow so many variants, there are so many ways to do something and the specification is huge.

Reply:
to get both 1. and 2. we need a class for each condition and a method to install them to the addressspace (in this case we need to use the node id) at the end of install we call init method that would be 1. 
the other way (2.) we only call init and give them an condition id
the key is here to have all functionality in a condition-type-class and make the addressspace implementation optional

Reply:
@Koseng Afaik: 1) is clear I guess, with every Node has it's own instance. 
In version 2) the ConditionSource (example: LevelMeasurement in a Tank) reference to the AlarmType. Via the _SourceNode Property_ (known from the BaseEventType) the _ConditionSource_ is known as well and an Alarm is triggered. So many Nodes may share the same Alarm, is that correct?

If yes, 2) may realy be a solution to spare and hide extra Nodes, even if 1) is more intuitive (in my opinion). We still have to deal with both, since Variant 3) (HasCondition used with an instance declaration) is wanted as well.

So it isn't necessarry to add an additional ConditionId, just the HasCondition Reference and the SourceNodeProperty attribute of the Event is needed?!

And if you want to use the condition in special domains, the ConditonClassId and ConditionSubClassId are useable.

Reply:
[https://reference.opcfoundation.org/v104/Core/docs/Part9/4.2](https://reference.opcfoundation.org/v104/Core/docs/Part9/4.2)
“Condition instances are specific implementations of a ConditionType. It is up to the Server whether such instances are also exposed in the Server’s AddressSpace. ”

I discussed the topic with Mr. Damm. Why are there two ways in the specification, one exposing the Condition-Instances in the AddressSpace, and the other way without exposing them? He said the reason is, because there are systems like the SIMATIC S7 PLC, where the server just doesn't know which relevant alarms/conditions for the system do exist. There are thousands of different alarms, created on the fly, even by different subsystems. Sometimes alarms are also dynamically build together. 

So, if you are in a scenario where you have a couple of alarms/conditions, which you know exactly, you should create the condition instances on startup and expose them in the AddressSpace. That is more common in the process industry, for example for some alarms of a tank. If possible, this is the preferred way, since the client can follow the OPC UA philosophy of browsing the AddressSpace and the client can find out which conditions do exist.

If you are in a scenario with a generic PLC like the SIMATIC S7 or with some controls of machine tools, you need to handle the Condition-Instances in a very dynamic way, since you don't know the conditions beforehand. For that scenario, you don't expose the condition instances in the AddressSpace. If conditions occur, only then the instances are created on the server dynamically. If the alarm/condition disappears, you can get rid of the instance as well. Also the ConditionId=NodeId is created dynamically, just for the lifetime of the condition instance. Usually then, many/all alarms are of the same type.

Then I asked, how to call a method, for example on an instance of AcknowledgeableConditionType, if the instance is not exposed in the AddressSpace? Because in my mind you always need a node from the addressSpace, where you can call the method on. He said that is not correct. A method call has 2 parameters. An ObjectId and a MethodId. In that case the client can execute a method call using the ConditionId as ObjectId.

So in the case of the python SDK, the client would directly call `async def call_method(parent, methodid, *args)` in `methods.py` without using a node object. “Parent” would be the ConditionId. From that perspective, the parameter name “parent” can be a bit misleading. It is just the identifier for the object you call the method on. See implementation in `async def _call_method(server, parentnodeid, methodid, arguments):` where the parameters for the request are called ObjectId and MethodId.

@swamper123 
“So many Nodes may share the same Alarm, is that correct?”
I don't exactly understand what you mean by that. 

In the case of not exposing the conditions in the AddressSpace, you just have a node where you can register for all the alarm/condition events. For example on a SIMATIC S7 you could have a node “ControlAlarms” where you subscribe for the AlarmConditionType-Events. The only purpose of the node is that there is a node where you can subscribe to the events of that type. Then if different alarms occurs, the alarm instances are created dynamically on the server side without exposing in the addressspace and the events are sent. Which each event you get the ConditionId=NodeId to keep track of existing alarms. Also the client can use the ConditionId to call a method on the internal condition instance.

Reply:
So to summarize:
- You need to create Condition-Instances on the server. The unique id of a Condition-Instance is the ConditionId of type NodeId.
- There are scenarios where the Condition-Instances are created and removed dynamically and not exposed in the AddressSpace
- There are scenarios where Condition-Instances are created static on startup and exposed via the AddressSpace.

Reply:
tbh have never seen a system which dont expose its alarms... is there a demoserver out there?
S7-1500 only support the dataaccess (read/write/subscribe/methods and since fw2.6 client functionality) maybe he spoilt siemens dev status ^^ 

i agree with the philosophy part but i also understand the usecase not exposing that alarm...

Reply:
> In the case of not exposing the conditions in the AddressSpace, you just have a node where you can register for all the alarm/condition events.

That was the answer to my question, thanks.

Reply:
@AndreasHeine I'll sent you an email.